processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,169 @@
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.metamx.common.ISE;
+import com.metamx.common.lifecycle.LifecycleStart;
+import com.metamx.common.lifecycle.LifecycleStop;
+import com.metamx.common.logger.Logger;
+import io.druid.guice.ManageLifecycle;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class provide a basic {@link LookupExtractorFactory} references manager.
+ * It allows basic operations fetching, listing, adding and deleting of {@link LookupExtractor} objects
+ * It is be used by queries to fetch the lookup reference.
+ * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP or other protocols.
+ */
+
+@ManageLifecycle
+public class LookupReferencesManager
+{
+  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
+  private final ConcurrentMap<String, LookupExtractorFactory> lookupMap = new ConcurrentHashMap();
+  private final Object lock = new Object();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lock) {
+      if (!started.getAndSet(true)) {
+        LOGGER.info("Started lookup factory references manager");
+      }
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (started.getAndSet(false)) {
+        LOGGER.info("Stopped lookup factory references manager");
+        for (String lookupName : lookupMap.keySet()) {
+          remove(lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName             namespace of the lookupExtractorFactory object
+   * @param lookupExtractorFactory {@link LookupExtractorFactory} implementation reference.
+   *
+   * @return true if the lookup is added otherwise false.
+   *
+   * @throws ISE If the manager is closed or if start of lookup returns false.
+   */
+  public boolean put(String lookupName, final LookupExtractorFactory lookupExtractorFactory) throws ISE
```

Comment:
`throws ISE` is not really needed because it is a runtime exception.  If you really want to have it there for documentation, please make it `throws IllegalStateException` instead.